### PR TITLE
Fix recursion error in packet list #1374

### LINKED ISSUE #1374: PacketList misses proper pickle support
#### Brief description

Pickling & unpickling PacketList causes RecursionError.

[Error initially reported](https://github.com/secdev/scapy/issues/1369) by @zlatdev (but closed cause of no response/further explanation).

#### Environment

- Scapy version: 2.4.0, 927c157b
- Python version: 3.5
- Operating System: Debian 9.3

#### How to reproduce

```
import pickle

frm = Ether(src='00:11:22:33:44:55', dst='00:11:22:33:44:55')/Raw()
pl = PacketList(res=[frm, frm])
pickled = pickle.dumps(pl)
pl = pickle.loads(pickled)
pl.show()
```

#### Actual result

```
...
RecursionError: maximum recursion depth exceeded
```

#### Expected result

Should show single frame like
```
0000 Ether / IP / TCP 127.0.0.1:ftp_data > 127.0.0.1:http S
```

*Edit: changed bug description as it turned out the issue is related to pickling (used by the SimpleQueue used by the initial error reporter)*

### PR DESCRIPTION
**Pull request by speakinghedge**
Allows pickling and unpickling of PacketList so it can be used with other containers like multiprocessing.SimpleQueue() that utilizes pickle to handle that queued data...

fixes #1374

### CODE REVIEW

File: scapy/plist.py [hunk 1]
```diff
@@ -85,6 +85,19 @@ def __repr__(self):
                                s,
                                ct.punct(">"))
 
+    def __getstate__(self):
```
Comment (guedou) [inline scapy/plist.py]: Could you add a doctring?

File: scapy/plist.py [hunk 2]
```diff
@@ -85,6 +85,19 @@ def __repr__(self):
...
+        state = {
+            'res': self.res,
+            'stats': self.stats,
+            'listname': self.listname
+        }
+        return state
+
+    def __setstate__(self, state):
```
Comment (guedou) [inline scapy/plist.py]: Could you add a doctring?

File: test/regression.uts [hunk 3]
```diff
@@ -10160,6 +10160,17 @@ if PYX:
...
 
+= __getstate__ / __setstate__ (used by pickle)
+
+import pickle
+
+frm = Ether(src='00:11:22:33:44:55', dst='00:11:22:33:44:55')/Raw()
+pl = PacketList(res=[frm, frm])
+pickled = pickle.dumps(pl)
+pl = pickle.loads(pickled)
+assert(len(pl) == 2)
```
Comment (gpotter2) [inline test/regression.uts]: how about:
```
assert(pl[0][Ether].src='00:11:22:33:44:55')
assert(pl[0][Ether].dst='00:11:22:33:44:55')
assert(pl[0].listname == "WhatAGreatName")
```

File: scapy/plist.py (hunk 1 above)
Comment (speakinghedge) [reply scapy/plist.py]: Yeah, I was thinking about that... and than I forgot about that... And you are right.

File: test/regression.uts [hunk 4]
```diff
@@ -10160,6 +10160,17 @@ if PYX:
     assert(os.path.exists(filename))
     os.unlink(filename)
 
+= __getstate__ / __setstate__ (used by pickle)
+
+import pickle
+
+frm = Ether(src='00:11:22:33:44:55', dst='00:11:22:33:44:55')/Raw()
+pl = PacketList(res=[frm, frm])
```
[the remark this answers is not in the document]
Comment (speakinghedge) [reply test/regression.uts]: You are right too. A missed chance to test something that is available is a broken test. Will fix later today,

[the remark this answers is not in the document]
Comment (speakinghedge) [reply test/regression.uts]: done in 9d478ab

**Comment by speakinghedge** [comment]
Thank you both for the prompt response and the valuable feedback. As a (sometimes) lazy person I need that kind of a trigger to complete my work. This could be the reason why I prefer pair programming...

**Pull request merged**